reference/options.md
# checked Options

Internally used, package-specific options. All options will prioritize R
options() values, and fall back to environment variables if undefined.
If neither the option nor the environment variable is set, a default
value is used.

## Checking Option Values

Option values specific to `checked` can be accessed by passing the
package name to `env`.

    options::opts(env = "checked")

    options::opt(x, default, env = "checked")

## Options

- tty_tick_interval:

  default:

  :   0.1

  option:

  :   checked.tty_tick_interval

  envvar:

  :   R_CHECKED_TTY_TICK_INTERVAL (evaluated if possible, raw string
      otherwise)

- tty_default_height:

  default:

  :   50

  option:

  :   checked.tty_default_height

  envvar:

  :   R_CHECKED_TTY_DEFAULT_HEIGHT (evaluated if possible, raw string
      otherwise)

- proactive_gc:

  default:

  :   TRUE

  option:

  :   checked.proactive_gc

  envvar:

  :   R_CHECKED_PROACTIVE_GC (evaluated if possible, raw string
      otherwise)

- results_error_on:

  default:

  :   "never"

  option:

  :   checked.results_error_on

  envvar:

  :   R_CHECKED_RESULTS_ERROR_ON (evaluated if possible, raw string
      otherwise)

- results_keep:

  default:

  :   "all"

  option:

  :   checked.results_keep

  envvar:

  :   R_CHECKED_RESULTS_KEEP (evaluated if possible, raw string
      otherwise)

- restore:

  default:

  :   NA

  option:

  :   checked.restore

  envvar:

  :   R_CHECKED_RESTORE (evaluated if possible, raw string otherwise)

- add_remotes:

  default:

  :   TRUE

  option:

  :   checked.add_remotes

  envvar:

  :   R_CHECKED_ADD_REMOTES (evaluated if possible, raw string
      otherwise)

- check_envvars:

  default:

  :   c(`_R_CHECK_FORCE_SUGGESTS_` = "false", `_R_CHECK_RD_XREFS_` = "false",
              `_R_CHECK_SYSTEM_CLOCK_` = "false", `_R_CHECK_SUGGESTS_ONLY_` = "true",
              `_R_CHECK_CRAN_INCOMING_` = "false")

  option:

  :   checked.check_envvars

  envvar:

  :   R_CHECKED_CHECK_ENVVARS (evaluated if possible, raw string
      otherwise)

- check_build_args:

  default:

  :   c("--no-build-vignettes", "--no-manual")

  option:

  :   checked.check_build_args

  envvar:

  :   R_CHECKED_CHECK_BUILD_ARGS (space-separated R CMD build flags)

- check_args:

  default:

  :   c("--timings", "--ignore-vignettes", "--no-manual", "--as-cran")

  option:

  :   checked.check_args

  envvar:

  :   R_CHECKED_CHECK_ARGS (space-separated R CMD check flags)

- install_envvars:

  default:

  :   callr::rcmd_safe_env()

  option:

  :   checked.install_envvars

  envvar:

  :   R_CHECKED_INSTALL_ENVVARS (evaluated if possible, raw string
      otherwise)

- install_system_profile:

  default:

  :   FALSE

  option:

  :   checked.install_system_profile

  envvar:

  :   R_CHECKED_INSTALL_SYSTEM_PROFILE (evaluated if possible, raw
      string otherwise)

- install_user_profile:

  default:

  :   "project"

  option:

  :   checked.install_user_profile

  envvar:

  :   R_CHECKED_INSTALL_USER_PROFILE (evaluated if possible, raw string
      otherwise)

- install_opts_to_inherit:

  default:

  :   list("timeout", "available_packages_filters", "HTTPUserAgent",
              "pkgType")

  option:

  :   checked.install_opts_to_inherit

  envvar:

  :   R_CHECKED_INSTALL_OPTS_TO_INHERIT (evaluated if possible, raw
      string otherwise)

## See also

options getOption Sys.setenv Sys.getenv

Other documentation:
[`options_params`](https://Genentech.github.io/checked/reference/options_params.md)


reference/options_params.md
# Checked Options

Checked Options

## Arguments

- proactive_gc:

  `logical`, indicating whether additional garbage collection should be
  performed before starting a new task, if at least one process recently
  finalized. This can cause the checker to orchestrate tasks slower but
  is recommended to be used for designs with many sub-processes required
  as native garbage collection can lag leading to memory issues. Disable
  only when maximum prefromance is required and memory is not the issue.
  (Defaults to `TRUE`, overwritable using option 'checked.proactive_gc'
  or environment variable 'R_CHECKED_PROACTIVE_GC')

- results_error_on:

  character vector indicating whether R error should be thrown when
  issues are discovered when generating results. "never" means that no
  errors are thrown. If "issues" then errors are emitted only on issues,
  whereas "potential issues" stands for error on both issues and
  potential issues. (Defaults to `"never"`, overwritable using option
  'checked.results_error_on' or environment variable
  'R_CHECKED_RESULTS_ERROR_ON')

- install_envvars:

  named `character` vector of environment variables to use during the
  package installation. (Defaults to
  [`callr::rcmd_safe_env()`](https://callr.r-lib.org/reference/rcmd_safe_env.html),
  overwritable using option 'checked.install_envvars' or environment
  variable 'R_CHECKED_INSTALL_ENVVARS')

- check_args:

  `character` vector of args passed to the R CMD check. (Defaults to
  `c("--timings", "--ignore-vignettes", "--no-manual", "--as-cran")`,
  overwritable using option 'checked.check_args' or environment variable
  'R_CHECKED_CHECK_ARGS')

- install_system_profile:

  `logical` used as `sytem_profile` parameter passed to the
  [`callr::r_bg()`](https://callr.r-lib.org/reference/r_bg.html)
  function used to install packages (Defaults to `FALSE`, overwritable
  using option 'checked.install_system_profile' or environment variable
  'R_CHECKED_INSTALL_SYSTEM_PROFILE')

- results_keep:

  character vector indicating which packages should be included in the
  results. "all" means that all packages are kept. If "issues" then only
  packages with issues identified, whereas "potential_issues" stands for
  keeping packages with both "issues" and "potential_issues". (Defaults
  to `"all"`, overwritable using option 'checked.results_keep' or
  environment variable 'R_CHECKED_RESULTS_KEEP')

- add_remotes:

  `logical` indicating whether origins inheriting from
  `pkg_origin_local`, should be scanned for packages in the `remotes`
  field and added while constrocuting a plan `task_grap` (Defaults to
  `TRUE`, overwritable using option 'checked.add_remotes' or environment
  variable 'R_CHECKED_ADD_REMOTES')

- check_envvars:

  named `character` vector of environment variables to use during the R
  CMD check. (Defaults to
  `c(`*R_CHECK_FORCE_SUGGESTS*`= "false",`*R_CHECK_RD_XREFS*`= "false", ; `*R_CHECK_SYSTEM_CLOCK*`= "false",`*R_CHECK_SUGGESTS_ONLY*`= "true", ; `*R_CHECK_CRAN_INCOMING*` = "false")`,
  overwritable using option 'checked.check_envvars' or environment
  variable 'R_CHECKED_CHECK_ENVVARS')

- install_opts_to_inherit:

  `list` of characters which specifies names of the R session options
  that should be inherited by the install subprocesses, from the main
  process (Defaults to
  `list("timeout", "available_packages_filters", "HTTPUserAgent", ; "pkgType")`,
  overwritable using option 'checked.install_opts_to_inherit' or
  environment variable 'R_CHECKED_INSTALL_OPTS_TO_INHERIT')

- tty_tick_interval:

  tty refresh interval when reporting results in milliseconds (Defaults
  to `0.1`, overwritable using option 'checked.tty_tick_interval' or
  environment variable 'R_CHECKED_TTY_TICK_INTERVAL')

- check_build_args:

  `character` vector of args passed to the R CMD build. (Defaults to
  `c("--no-build-vignettes", "--no-manual")`, overwritable using option
  'checked.check_build_args' or environment variable
  'R_CHECKED_CHECK_BUILD_ARGS')

- restore:

  `logical` indicating whether output directory should be unlinked
  before running checks. If `FALSE`, an attempt will me made to restore
  previous progress from the same `output` (Defaults to `NA`,
  overwritable using option 'checked.restore' or environment variable
  'R_CHECKED_RESTORE')

- install_user_profile:

  value used as `user_profile` parameter passed to the
  [`callr::r_bg()`](https://callr.r-lib.org/reference/r_bg.html)
  function used to install packages (Defaults to `"project"`,
  overwritable using option 'checked.install_user_profile' or
  environment variable 'R_CHECKED_INSTALL_USER_PROFILE')

- tty_default_height:

  deafult tty height used for the ANSI reporter. Used only if correct
  values could not be acquired with system('tput lines') (Defaults to
  `50`, overwritable using option 'checked.tty_default_height' or
  environment variable 'R_CHECKED_TTY_DEFAULT_HEIGHT')

## See also

Other documentation:
[`options()`](https://Genentech.github.io/checked/reference/options.md)


NEWS.md
# checked 0.5.5

* Add 'install_opts_to_inherit' option which can be used to specify parameters
  that install subprocess should inherit from the main process

# checked 0.5.4

* Improve error messaging when using basic_tty

* Various improvements to reporters and minor display tweaks

# checked 0.5.3

* Make is possible to construct reporter environments with additional values
  to better control their behavior.

* Add `checks_only` parameter to `reporter_basic_tty` which if set to TRUE
  will make the reporter broadcast only check tasks instead of both
  install and check.

* Address changes related to default parameters changes in `callr`
  and resulting NULL comparison in `checked` (@gaborcsardi, #93)
  
* Multiple API changes facilitating additional customization for tasks. 

* Ensure packages destined into isolated libraries or those coming from
  non-standard sources are always installed.

* Redesign logs by grouping them into package specific directories.

# checked 0.5.2

* Add timers striping to `strip_details_from_issue()` to avoid false-positives.

* Remove `pkg_origin_is_base()` helper function and use memoised `base_pkgs()`.

* Update `RE_CHECK` to capture even more edge cases while parsing R CMD check
  output.

* Finish check process even if checks seem incomplete but 3 or mire minutes have
  passed since the process ended to avoid infinite loops.

* Further improvements to the check process finisher.

* Make `graph_dedup_attrs` rebuild the graph from scratch with deduplicated
  attributes rather than manipulating the exiting graph. I significantly speeds
  up the function.

# checked 0.5.1

* Export STATUS vector to make external statuses analysis easier.

* Allow obtaining partial results for uncompleted checks.

* Export `results_to_df()` and make it a generic.

# checked 0.5.0

* Entire package refactor

# checked 0.2.3

* Fix check processes hanging forever in some system configurations.

* Remove all whitespaces (`"[[:space:]]"`) before comparing potential issues
  to skip most of the false positives.

* Shortened many user-facing function names. In general, uses of `reverse`
  were shortened to `rev`, `dependencies` was shortened to `deps`,
  `development` was shortened to `dev` and `package` was shortened to `pkg`.
  For example, `check_reverse_dependencies_development()` was shortened to
  `check_dev_rev_deps()` (@dgkf-roche)

* Fix names in the `enum` function for R lower than 4.3.

* Add tests for the reverse dependency check use case.

* Make `install_process` capture session's `available_packages_filters`
  and reuse them when installing packages to ensure consistency with the main
  session filtering.

* Prettify output by stripping excessive new lines.

* Force garbage collection before scheduling task, to make sure any already
  finished processes are removed from the memory.

# checked 0.2.3

* Use custom `checked` `finisher`'s instead of the `processx` `finalizer`'s
  when cleaning up finished processes to avoid executing callbacks when
  objects are garbage collected.
  
* In `rev_dep_check_tasks_df` add custom package value for each check associated
  with the release version of the package to make sure, the current release
  version of the package is always fetched and installed.
  
* Prevent `results` from attempting to read results for unfinished checks.

# checked 0.2.2

* Clear some of the unused utils functions (`can_symlink`, `symlink_or_copy`).

* Make sure output directory is always created.

# checked 0.2.1 

* Add `results_to_file` function.

# checked 0.2.0 

* Change default private `lib.loc` value in `check_process` R6 class
  to NULL to avoid staged installation failures.
  
* Terminate all the subprocesses if the `run` function is interrupted to
  comply with CRAN requirements.
  
* Reorganize methods in the `check_design` making some of the private.

* Remove warning about possible problems with isolation.

* Rename `development_only` parameter to `versions` in the `rev_dep_check_tasks_df`

* Make sure `check_reverse_dependencies` works when package is not available
  in the remote sources or has no reverse dependencies.
  
* Add `keep` parameters to the the `print.checked_results_check_task_spec`.

* Add `error_on` parameters to the the `results.check_design`.

* Add `$get_r_exit_status()` method to both check and install processes.

* Add warnings to the reporter whenever sub-processes have non-zero exit status.

* Fix race condition when reporting progress for check processes.

# checked 0.1.0

* Package released to CRAN

R/options.R
#' @import options
options::define_options(
  "tty refresh interval when reporting results in milliseconds",
  tty_tick_interval = 0.1,

  "deafult tty height used for the ANSI reporter. Used only
   if correct values could not be acquired with system('tput lines')",
  tty_default_height = 50,

  "`logical`, indicating whether additional garbage collection should be
   performed before starting a new task, if at least one process recently
   finalized. This can cause the checker to orchestrate tasks slower but
   is recommended to be used for designs with many sub-processes required as
   native garbage collection can lag leading to memory issues. Disable only
   when maximum prefromance is required and memory is not the issue.",
  proactive_gc = TRUE,

  "character vector indicating whether R error should be thrown when issues
   are discovered when generating results. \"never\" means that no errors
   are thrown. If \"issues\" then errors are emitted only on issues, whereas
   \"potential issues\" stands for error on both issues and potential issues.",
  results_error_on = "never",

  "character vector indicating which packages should be included in the results.
   \"all\" means that all packages are kept. If \"issues\" then only packages 
   with issues identified, whereas \"potential_issues\" stands for keeping
   packages with both \"issues\" and \"potential_issues\".",
  results_keep = "all",

  "`logical` indicating whether output directory should be unlinked before
   running checks. If `FALSE`, an attempt will me made to restore previous
   progress from the same `output`",
  restore = NA,

  "`logical` indicating whether origins inheriting from `pkg_origin_local`, 
    should be scanned for packages in the `remotes` field and added while 
    constrocuting a plan `task_grap`",
  add_remotes = TRUE,

  "named `character` vector of environment variables to use during
   the R CMD check.",
  check_envvars = c(
    "_R_CHECK_FORCE_SUGGESTS_" = "false",
    "_R_CHECK_RD_XREFS_" = "false",
    "_R_CHECK_SYSTEM_CLOCK_" = "false",
    "_R_CHECK_SUGGESTS_ONLY_" = "true",
    "_R_CHECK_CRAN_INCOMING_" = "false"
  ),

  "`character` vector of args passed to the R CMD build.",
  check_build_args = c(
    "--no-build-vignettes",
    "--no-manual"
  ),
  envvar_fn = structure(
    function(raw, ...) trimws(strsplit(raw, " ")[[1]]),
    desc = "space-separated R CMD build flags"
  ),

  "`character` vector of args passed to the R CMD check.",
  check_args = c(
    "--timings",
    "--ignore-vignettes",
    "--no-manual",
    "--as-cran"
  ),
  envvar_fn = structure(
    function(raw, ...) trimws(strsplit(raw, " ")[[1]]),
    desc = "space-separated R CMD check flags"
  ),

  "named `character` vector of environment variables to use during
   the package installation.",
  install_envvars = callr::rcmd_safe_env(),

  "`logical` used as `sytem_profile` parameter passed to the `callr::r_bg()`
   function used to install packages",
  install_system_profile = FALSE,

  "value used as `user_profile` parameter passed to the `callr::r_bg()`
   function used to install packages",
  install_user_profile = "project",

  "`list` of characters which specifies names of the R session options that
   should be inherited by the install subprocesses, from the main process ",
  install_opts_to_inherit = list(
    "timeout",
    "available_packages_filters",
    "HTTPUserAgent",
    "pkgType"
  )
)

#' @eval options::as_roxygen_docs()
#'
#' @family documentation
NULL

#' Checked Options
#' @eval options::as_params()
#' @name options_params
#'
#' @family documentation
NULL
